pkg/gateway/gateway.go

```diff
@@ -208,3 +207,21 @@ func (g *gateway) Run(ctx context.Context, done chan<- struct{}) {
 
 	close(done)
 }
+
+func (g *gateway) updateGateway(doc *api.GatewayDocument) {
+
+	if doc.Gateway.Deleting {
+		// https://docs.microsoft.com/en-us/azure/cosmos-db/change-feed-design-patterns#deletes
+		delete(g.gateways, doc.ID)
+	} else {
+		g.gateways[doc.ID] = doc.Gateway
+	}
+}
+
+func (g *gateway) updateGateways(docs []*api.GatewayDocument) {
+	g.mu.Lock()
+	defer g.mu.Unlock()
```

Comment:
Yeah I was hesitating between the two and decided to keep it there because I wanted to keep the exact same logic. But I think you're right actually.

Reply:
now the question becomes: is there a reason we used to block for the whole set, other than saving a few CPU cycles? 🤔

Reply:
I think not. The whole thing is not deterministic and I think that it's only eventually consistent with cosmosdb (because of the way it is synchronized with cosmosdb). The worst case is that the next change will be picked up at the next iteration (after the ticker ticks), but that could have happened with the more aggressive lock too.